### ISSUE TITLE
#positive_definite?, #positive_semidefinite?, #rank_deficient?

### ISSUE THREAD

**Issue by mohawkjohn** [SciRuby/nmatrix#411]
These methods, among others, would be useful for users wanting to check matrix properties.

**Comment by johnmarinelli** [comment]
Hi!  I'd like to take a shot at this.  I think I can have something to show within a week or two.

**Comment by mohawkjohn** [comment]
Wonderful! Thanks so much for getting involved. =D

**Comment by johnmarinelli** [comment]
@mohawkjohn Hey!  I'm afraid I don't understand what the second parameter of NMatrix::rank is supposed to do, but if I can use that function to calculate the rank of the matrix then rank_deficient? would be really easy to implement.

**Comment by mohawkjohn** [comment]
`#rank` basically allows you to iterate along either rows or columns or through the depth of the tensor if it's more than 2-dimensional. So the first argument is row, column, or depth; and the second argument is which one of those. If you do `rank(0,1)`, it gives you row 1. If you do `rank(1,5)`, it gives you column 5. If you do `rank(2,1)`, you get layer 1. Make sense?

**Comment by johnmarinelli** [comment]
@mohawkjohn yep, perfect sense!  I ended up using the gesvd to calculate rank.  Since this needs the nmatrix/lapacke gem, I have a couple of questions -
1) Is that okay?  Or did you want it to be implemented without the lapacke dependency?
2) How can I test the nmatrix functions that require lapacke?  The tests that are already in place give me a NotImplementedException, even though I have the nmatrix-lapacke gem installed.

**Comment by wlevine** [comment]
You can run the tests with `rake spec nmatrix_plugins=lapacke` which will
require the nmatrix/lapacke gem

**Comment by mohawkjohn** [comment]
Is using gesvd the most economical way to calculate rank?

**Comment by johnmarinelli** [comment]
@mohawkjohn , well, I know there are a variety of ways to calculate whether or not a matrix is full rank or not; but I thought that SVD was the only way to calculate rank?

**Comment by mohawkjohn** [comment]
Yup, you're right. Had to check my reference. =) Yes, lapacke sounds good.

**Comment by johnmarinelli** [comment]
@mohawkjohn sounds good; i will continue to push on :) @wlevine thanks!
